howso/residuals.amlg

```diff
@@ -129,62 +181,127 @@
-										(map
-											(lambda (let
-												(assoc class_prob (* (/ (current_value 1) total_count)) )
-												(* class_prob (- 1 class_prob))
+											(compute_on_contained_entities (list
+												(query_value_masses current_feature (null) (contains_index !numericNominalFeaturesMap current_feature) )
 											))
-											(values class_counts_map)
 										)
-									)
+								))
 
+								;do not max cap nominal deviations if less than 2 classes have been trained
+								(if (<= (size class_counts_map) 1)
 									(max
 										(get !cachedFeatureMinResidualMap (current_index))
 										(current_value)
 										(get !userSpecifiedFeatureErrorsMap (current_index))
 									)
+
+									;else cap the max nominal deviation
+									(let
+										(assoc total_count (apply "+" (values class_counts_map)) )
+
+										(min
+											;nominal max deviation is the sum of: each class's probability multiplied by probability of getting it wrong
+											(apply "+"
+												(map
+													(lambda (let
+														(assoc class_prob (* (/ (current_value 1) total_count)) )
+														(* class_prob (- 1 class_prob))
+													))
+													(values class_counts_map)
+												)
+											)
+
+											(max
+												(get !cachedFeatureMinResidualMap (current_index))
+												(current_value)
+												(get !userSpecifiedFeatureErrorsMap (current_index))
+											)
+										)
+									)
 								)
 							)
+
+							;else continuous value don't have upper bounds, set the lower bound
+							(max
+								(get !cachedFeatureMinResidualMap (current_index))
+								(current_value)
+								(get !userSpecifiedFeatureErrorsMap (current_index))
+							)
 						)
 					)
 
-					;else continuous value don't have upper bounds, set the lower bound
-					(max
-						(get !cachedFeatureMinResidualMap (current_index))
-						(current_value)
-						(get !userSpecifiedFeatureErrorsMap (current_index))
+					;if using shared deviations, map over a reduced residuals map with only features that are not
+					; in a shared deviations group or are the primary keys of a shared deviations group
+					(if (size !sharedDeviationsMap)
+						(call !ReduceSharedDeviationsMapping (assoc map_to_reduce (get feature_residuals_map "residual_map") ))
+						(get feature_residuals_map "residual_map")
 					)
 				)
 			)
-			(get feature_residuals_map "residual_map")
+
+			(if (size !sharedDeviationsMap)
+				(call !ExpandForSharedDeviations (assoc
+					compressed_values temp_residual_map
+				))
+				temp_residual_map
+			)
+		)
+	)
+
+
+	;Takes a assoc of of values and filters the assoc to only features that either do not belong to a group of shared deviations, or are the primary
+	; key for a group of shared deviation features.
+	;
+	;parameters:
+	; map_to_reduce: An assoc that is to be reduced.
+	#!ReduceSharedDeviationsMapping
+	(declare
+		(assoc map_to_reduce (null))
+		(let
```

Comment:
This whole method can be simplified into:
`(remove map_to_reduce
    ;remove non-primary key features
    (remove 
        (indices !sharedDeviationsMap)
        ;primary keys
        (indices (zip (values !sharedDeviationsMap))
    )
)`

Though since this pattern, as well as just having primary keys is going to be used over and over, it may make sense to cache them into model variables as well. perhaps !sharedDeviationsPrimaryFeatures and !sharedDeviationsNonPrimaryFeatures ? That we there's no need to do `(indices (zip (values !sharedDeviationsMap)))` every time you need access to just the primary features